DIPs/1NNN-RAC.md

```diff
@@ -0,0 +1,333 @@
+
+Different strategies for describing how to select DllImport and internal modes, along with DllExport is presented next. These are:
+
+- Positive notation using ``export``. This is the default behavior of a D compiler.
+- Negative notation using visibility override switch. By default all symbols are not exported. You can use the visibility override switch to force all symbols to be exported instead. This allows for libraries that are not currently marked with ``export`` to export their symbols.
+- Lastly a corner case notation for when you need to determine if a symbol is internal or DllImport depending on build step. Useful for multi-step builds where interesting or conflicting symbols modes are present.
+
+#### Export Annotation
+
+The ``export`` attribute is given an identifier parameter which is interpreted to be a version.
+
+```diff
+VisibilityAttribute:
+-   export
+
+Attribute:
++   export
++   export ( Identifier )
+```
+
+This parameter when the identifier is active (specified by ``-version=ident``) places the symbol when not compiling into internal mode.
+
+When not active it places it into DllImport mode as if ``extern`` was also annotated.
+
+To standardise the identifiers in use, three new prefixes are added to the specification, each will have instances for libc, druntime and Phobos that will be provided by the compiler automatically.
+
+The prefixes are ``Have_``, ``InBinary_``, and ``Compiling_``. The suffixes for these three will be a logical package.
+
+- ``Have_`` is used when a given logical package is available as a dependency during linking.
+- ``InBinary_`` is used when a given logical package has its symbols in the currently compiling binary.
+- ``Compiling_`` is used when a given logical package is currently being compiled.
+
+Between these three versions it is possible to cover any corner case where a symbol must be viewed internally instead of out of binary.
+
+The D interface generator will need to support the insertion of the version argument ``InBinary_``. This will be needed to accurately represent C files by a D module. The mechanism to specify what the suffix is, is not determined here. It may depend upon C preprocessor capability.
+
+#### Positive Notation
+
+To assign a symbol that it is exported, use the ``export`` attribute to annotate a D symbol as DllExport.
+
+The export annotation must not be a visibility modifier.
+If it is a visibility modifier you are required to expose internal implementation details that could result in unsafe operation of a codebase by external parties at the language level.
```

Comment:
Examples/justification?

Reply:
Here is some real code exemplifying it:

https://github.com/Project-Sidero/basic_memory/blob/36498e45cb2f9a8cdafaee099a7beb2ba233416f/source/sidero/base/path/file.d#L1012

If you don't do this, you expose internals.

Exportation is fundamentally a linker concept, not a language one. Tieing it to visibility results in making things that should be private public.

You recently did a talk saying not to do this, I would hope that this wouldn't need elaborating ;)